### PR TITLE
OCPERT-97 Add notify-managers command

### PR DESCRIPTION
**Pull request by tomasdavidorg**

### CODE REVIEW

**Review by LuboTerifaj — COMMENTED** [review]
Hello @tomasdavidorg ,

Although this is not strictly necessary, it might be worth adding checks similar to https://github.com/openshift/release-tests/blob/b28a7f6de2f341d3537770ddbf7278bd8ed0b920/oar/cli/cmd_update_bug_list.py#L23

Other than that, it looks good to me.

Thanks!

**Comment by rioliu-rh** [comment]
# Code Review Comments for PR #541

## General Observations
- The changes introduce new features and improvements to the bug tracking and notification system.
- The code modifications appear to be well-structured and follow existing patterns.
- New test files and helper classes are added, which is good for maintainability.

## Specific Comments

### oar/cli/cmd_update_bug_list.py
1. **Lines 16-19**: The new `notify-team-leads` and `notify-managers` options are well-documented with clear help messages.
2. **Lines 39-40**: The implementation of `notify_managers` calls `share_unverified_cve_issues_to_managers`, but the method is not yet fully implemented (line 39 has a warning).
3. **Lines 41-42**: The `notify_team_leads` option is not yet implemented (line 39 has a warning).

### oar/core/jira.py
1. **Lines 237-243**: The new `get_unverified_cve_issues` method is well-documented and appears correct.
2. **Lines 245-254**: Consider adding more detailed logging for debugging purposes when fetching unverified CVE issues.

### oar/core/ldap.py
1. **Lines 1-138**: The new `LdapHelper` class is well-structured. Consider adding more comprehensive error handling for connection failures.
2. **Lines 103-110**: The `_get_manager_id` method could benefit from additional logging to track the search process.
3. **Lines 140-153**: The `_get_user_email` method is well-implemented but could use more detailed logging for debugging.

### oar/core/notification.py
1. **Lines 268-278**: The new `share_unverified_cve_issues_to_managers` method is well-documented and appears correct.
2. **Lines 271-273**: Consider adding a fallback mechanism if the Slack channel lookup fails.

### pyproject.toml
1. **Lines 53-54**: The addition of `ldap3 >= 2.9.1` is appropriate given the new LDAP functionality.

### tests/test_ldap.py
1. **Lines 1-37**: The new test file is well-structured and covers the main functionality of the `LdapHelper` class.
2. **Lines 25-30**: The test for `get_manager_email_from_user_email` could include more edge cases.
3. **Lines 32-36**: The test for `get_group_members_emails` assumes a specific group structure. Consider adding comments explaining this assumption.

**Comment by rioliu-rh** [comment]
# Code Review Comments for PR #541

## General Observations
- The changes introduce new features and improvements to the bug tracking and notification system.
- The code modifications appear to be well-structured and follow existing patterns.
- New test files and helper classes are added, which is good for maintainability.

## Specific Comments

### oar/cli/cmd_update_bug_list.py
1. **Lines 16-19**: The new `notify-team-leads` and `notify-managers` options are well-documented with clear help messages.
2. **Lines 39-40**: The implementation of `notify_managers` calls `share_unverified_cve_issues_to_managers`, but the method is not yet fully implemented (line 39 has a warning).
3. **Lines 41-42**: The `notify_team_leads` option is not yet implemented (line 39 has a warning).

### oar/core/jira.py
1. **Lines 237-243**: The new `get_unverified_cve_issues` method is well-documented and appears correct.
2. **Lines 245-254**: Consider adding more detailed logging for debugging purposes when fetching unverified CVE issues.

### oar/core/ldap.py
1. **Lines 1-138**: The new `LdapHelper` class is well-structured. Consider adding more comprehensive error handling for connection failures.
2. **Lines 103-110**: The `_get_manager_id` method could benefit from additional logging to track the search process.
3. **Lines 140-153**: The `_get_user_email` method is well-implemented but could use more detailed logging for debugging.

### oar/core/notification.py
1. **Lines 268-278**: The new `share_unverified_cve_issues_to_managers` method is well-documented and appears correct.
2. **Lines 271-273**: Consider adding a fallback mechanism if the Slack channel lookup fails.

### pyproject.toml
1. **Lines 53-54**: The addition of `ldap3 >= 2.9.1` is appropriate given the new LDAP functionality.

### tests/test_ldap.py
1. **Lines 1-37**: The new test file is well-structured and covers the main functionality of the `LdapHelper` class.
2. **Lines 25-30**: The test for `get_manager_email_from_user_email` could include more edge cases.
3. **Lines 32-36**: The test for `get_group_members_emails` assumes a specific group structure. Consider adding comments explaining this assumption.

**Comment by rioliu-rh** [comment]
### Review Comments for PR #541

#### oar/cli/cmd_update_bug_list.py
- Lines 16-19: The new options `--notify-team-leads` and `--notify-managers` are well-documented with clear help messages.
- Line 39: The implementation for `notify_team_leads` is incomplete (placeholder warning). Consider marking this as a TODO or raising an error since it's not fully implemented.
- Lines 41-42: The `notify_managers` option correctly calls `share_unverified_cve_issues_to_managers`, but ensure this method is properly implemented and tested.

#### oar/core/jira.py
- Lines 237-249: The new `get_unverified_cve_issues` method is well-documented and correctly identifies issues that are both on QA and marked as CVE trackers. This implementation looks solid.

#### oar/core/notification.py
- Lines 268-279: The new `share_unverified_cve_issues_to_managers` method correctly integrates with the LDAP helper to notify managers of unverified CVE issues. Ensure that error handling is comprehensive and that the Slack channel is correctly configured.

#### oar/core/ldap.py
- Lines 1-138: The new `LdapHelper` class is well-structured with clear methods for:
  - Getting manager emails from user emails
  - Fetching group members' emails
  - Error handling for LDAP connection issues
  - Proper logging of errors
- Consider adding more comprehensive unit tests for the LDAP helper, especially in edge cases like invalid credentials or connection failures.

#### oar/tests/test_ldap.py
- Lines 1-37: The new test file `test_ldap.py` effectively tests the core functionality of `LdapHelper`, including:
  - Testing manager email retrieval
  - Testing group member email retrieval
  - Testing manager email retrieval from user emails
- Ensure that the test data used is representative of real-world scenarios and consider adding more test cases for edge cases.

#### General Comments:
- The addition of `ldap3 >= 2.9.1` to `pyproject.toml` is appropriate given the new LDAP functionality.
- The implementation of notifying managers for unverified CVE issues is a strong improvement for security tracking.
- Ensure that all new functionality is properly integrated into the existing workflow and that there are no breaking changes for existing users of the CLI tool.

#### Suggestions for Improvement:
- Consider adding more detailed logging in the `LdapHelper` methods to track the number of successful vs failed queries.
- Ensure that the placeholder warning for `notify_team_leads` is addressed in a follow-up PR to complete the feature.
- Consider adding rate limiting or retry logic for LDAP queries to handle transient connection issues.

**Comment by rioliu-rh** [comment]
# Code Review for PR #541

## Summary
This PR adds functionality to notify managers about unverified CVE issues via:
1. New CLI options `--notify-team-leads` and `--notify-managers`
2. New LDAP integration to get manager emails
3. New notification methods for unverified CVEs

## Potential Issues

### Unimplemented Feature
- The `--notify-team-leads` option is not yet implemented (just logs a warning)
  - Consider either implementing this or removing the option until ready

### LDAP Integration
- The LDAP server URL is hardcoded (`ldaps://ldap.corp.redhat.com`)
  - Should this be configurable?
- Error handling could be more robust for LDAP operations
- Test cases use real email addresses - consider using test/mock data

### Dependencies
- New dependency `ldap3` added to pyproject.toml
  - Should be documented in README/CHANGELOG
  - Need to verify version compatibility

### Testing
- Additional test cases could be added for:
  - Failed LDAP connections
  - Edge cases in manager/group lookups
  - Notification formatting

## Suggestions
- Consider making LDAP server URL configurable
- Implement or remove the team leads notification option
- Add more test coverage for error cases
- Document new dependency and features

**Comment by rioliu-rh** [comment]
## PR Review for openshift/release-tests#541

### Overview
This PR adds new functionality for notifications related to CVE issues, specifically adding the ability to notify team leads and managers about unverified CVE issues. It includes LDAP integration for manager lookups.

### Key Changes
1. New CLI options:
   - `--notify-team-leads`
   - `--notify-managers`

2. New Components:
   - Added new `LdapHelper` class for LDAP server integration
   - Added new methods in `JiraManager` for CVE issue handling
   - Extended notification capabilities in `NotificationManager`

### Potential Issues and Suggestions

#### Security & Configuration
1. LDAP connection is hardcoded to "ldaps://ldap.corp.redhat.com"
   - Consider making this configurable via configuration file
   - Ensure LDAPS (secure) is always used

#### Code Quality
1. The `notify-team-leads` option is added but noted as "not yet implemented"
   - Consider either implementing this feature or removing it until ready
   - Add TODO/FIXME comment with tracking issue if planned for future

2. Error Handling:
   - LDAP connection errors are logged but the methods return None silently
   - Consider propagating errors up or adding retry mechanisms
   - Add documentation about error scenarios

#### Testing
1. Unit Tests:
   - The LDAP tests include hardcoded email addresses and expected values
   - Consider using mock data or configuration for test cases
   - Add negative test cases for error scenarios

#### Dependencies
1. Added new dependency "ldap3 >= 2.9.1"
   - Ensure this is compatible with all deployment environments
   - Document any new system requirements (e.g., LDAP libraries)

### Recommendations
1. Add documentation for the new LDAP integration
2. Consider adding configuration options for LDAP connection parameters
3. Implement or remove the unimplemented `notify-team-leads` option
4. Enhance error handling for LDAP operations
5. Add more test coverage for error cases

Overall, the changes look well-structured but could benefit from additional error handling and configuration flexibility.
EOF 2>&1

**Comment by rioliu-rh** [comment]
Above 2 review comments are generated by AI.

File: oar/core/jira.py [hunk 1]
```diff
@@ -237,6 +237,23 @@ def get_high_severity_and_can_drop_issues(self, jira_issue_keys):
...
+    
+    def get_unverified_cve_issues(self, jira_issue_keys: list[str]):
+        """
+        Get list of unverified CVE issues
+
+        Args:
+            jira_issue_keys (list[str]): List of all jira issues keys
+
+        Returns:
+            List[JiraIssue]: List of all unverified CVE issues
```
[the remark this answers is not in the document]
Comment (tomasdavidorg) [reply oar/core/jira.py]: Fixed

File: oar/cli/cmd_update_bug_list.py [hunk 2]
```diff
@@ -16,12 +16,16 @@
...
+def update_bug_list(ctx, notify, confirm_droppable, notify_team_leads, notify_managers):
     """
     Update bug status listed in report, update existing bug status and append new ON_QA bug
     """
-    if not notify and confirm_droppable:
-        raise click.UsageError("Error: --no-notify and --confirm-droppable cannot be used together")
+    if not notify and (confirm_droppable or notify_team_leads or notify_managers):
+        raise click.UsageError("Error: --no-notify cannot be used together with --confirm-droppable, --notify-team-leads or --notify-managers")
+    if sum([confirm_droppable, notify_team_leads, notify_managers]) > 1:
+        raise click.UsageError("Error: only one of parameters --confirm-droppable, --notify-team-leads or --notify-managers can be used simultaneously")
```
Comment (tomasdavidorg) [inline oar/cli/cmd_update_bug_list.py:28]: @LuboTerifaj Please take a look. I updated the checks:

1. `--no-notify` cannot be used with any of `--confirm_droppable` or `--notify_team_leads` or `--notify_managers`
2. Only one of `--confirm_droppable` or `--notify_team_leads`, `--notify_managers` can be used simultaneously

**Comment by tomasdavidorg** [comment]
Comments to the PR reviews:

- Consider making LDAP server URL configurable / Consider adding configuration options for LDAP connection parameters
    - I don't think we need to configure the LDAP server. 
-  Implement or remove the team leads notification option / Implement or remove the unimplemented notify-team-leads option
    - I will do that
- Add more test coverage for error cases / Add more test coverage for error cases
    - I will check I can add some tests
- Document new dependency and features / Add documentation for the new LDAP integration
    - I haven't found any place where it could be documented. When the PR is merged I will update the corresponding documents in Google drive
-  Enhance error handling for LDAP operations
    - I will look at this and check if I can propagate the error

File: oar/cli/cmd_update_bug_list.py [hunk 3]
```diff
@@ -37,6 +40,9 @@ def update_bug_list(ctx, notify, confirm_droppable):
             if confirm_droppable:
                 high_severity_issues, _ = JiraManager(cs).get_high_severity_and_can_drop_issues(jira_issues)
                 NotificationManager(cs).share_high_severity_bugs(high_severity_issues)
+            elif notify_managers:
+                unverified_cve_issues = JiraManager(cs).get_unverified_cve_issues(jira_issues)
+                NotificationManager(cs).share_unverified_cve_issues_to_managers(unverified_cve_issues)
```
Comment (rioliu-rh) [inline oar/cli/cmd_update_bug_list.py:45]: suggest to check if issue list is empty

File: oar/core/ldap.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            user_email (str): User primary email
+
+        Returns:
+            str: Found manager email for specified user email
+        """
+
+        if user_email is None or not is_valid_email(user_email):
+            raise LdapHelperException(f"Specified email is not valid: {user_email}")
+
+        manager_id = self._get_manager_id(user_email)
```
Comment (rioliu-rh) [inline oar/core/ldap.py:39]: is it possible that `manager_id` is `None` here?

Comment (tomasdavidorg) [reply oar/core/ldap.py:35]: I added one more check to `_get_manager_id` and move the check from `get_manager_email` to `_get_manager_id`

File: oar/cli/cmd_update_bug_list.py (hunk 3 above)
Comment (tomasdavidorg) [reply oar/cli/cmd_update_bug_list.py:45]: This is covered here https://github.com/openshift/release-tests/pull/541/files#diff-a381d0d7c9b68a8747eb48b4124cdba8d709b5da46fb0b26d9e539a45b729e6cR544

File: oar/core/ldap.py (hunk 4 above)
Comment (rioliu-rh) [reply oar/core/ldap.py:35]: I want to confirm the ldap api behavior, if the input mail address is invalid (none, wrong mail id etc), what will happen, error will be raised or return none instead. current logic is two steps operation, we need to handle the case that manager id is none but no error raised.

File: oar/cli/cmd_update_bug_list.py (hunk 3 above)
Comment (rioliu-rh) [reply oar/cli/cmd_update_bug_list.py:45]: IMO, even notification api can handle this case, in this case to make the logic more clear and meaningful, we should not call notification func if there is no unverified cve issue found. implicit handling is not a good practice.

File: oar/core/ldap.py (hunk 4 above)
Comment (tomasdavidorg) [reply oar/core/ldap.py]: the `conn.entries` returns empty list.

Comment (rioliu-rh) [reply oar/core/ldap.py]: So it means, no ldap exception will be raised. the manager UID could be None, am I right?

Comment (tomasdavidorg) [reply oar/core/ldap.py]: Yes, there shouldn't be any exception

Comment (rioliu-rh) [reply oar/core/ldap.py]: ok, the error handling logic should be enhanced here, do validation for `manager_id` before calling `_get_user_email `, log warning message and return directly or raise exception. I prefer option#1, if there are multiple unverified cve issues, the process will not exit

File: oar/cli/cmd_update_bug_list.py (hunk 3 above)
Comment (tomasdavidorg) [reply oar/cli/cmd_update_bug_list.py]: I added the check for `unverified_cve_issues` and also for `high_severity_issues`. Notification methods won't be called if the lists are empty.

File: oar/core/ldap.py (hunk 4 above)
Comment (tomasdavidorg) [reply oar/core/ldap.py:32]: I fixed the class. Now the process is not interrupted. If a record is not found for specified parameters, `None` is returned.  In the notification class, warning is send:
`logger.warning(f"Manager email was not found for user {qa_contact_email}")`

**Comment by tomasdavidorg** [comment]
@rioliu-rh @LuboTerifaj I hope I resolve all the suggestions. Please take a look again.

**Comment by rioliu-rh** [comment]
lgtm
/assign @LuboTerifaj

**Review by LuboTerifaj — APPROVED** [review]
It looks good to me. Thanks @tomasdavidorg

File: oar/cli/cmd_update_bug_list.py (hunk 2 above)
Comment (LuboTerifaj) [reply oar/cli/cmd_update_bug_list.py]: Thanks!

**Comment by rioliu-rh** [comment]
@tomasdavidorg Can you cherry pick this change to branch `oar-errata-release-flow` manually? thanks

**Comment by tomasdavidorg** [comment]
> @tomasdavidorg Can you cherry pick this change to branch `oar-errata-release-flow` manually? thanks

Sure. No problem 🙂
